src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -1474,7 +1474,8 @@ private MethodSymbol FindEntryPoint(CancellationToken cancellationToken, out Imm
                     if (candidate.IsAsync)
                     {
                         // PROTOTYPE(async-main): Get the diagnostic to point to a smaller syntax piece.
-                        CheckFeatureAvailability(candidate.DeclaringSyntaxReferences.Single().GetSyntax(), MessageID.IDS_FeatureAsyncMain, diagnostics);
+                        // PROTOTYPE(async-main): Switch diagnostics around if the type is not Task or Task<T>
+                        CheckFeatureAvailability(candidate.ExtractReturnTypeSyntax(), MessageID.IDS_FeatureAsyncMain, diagnostics);
```

Comment:
>CheckFeatureAvailability(candidate.ExtractReturnTypeSyntax(), MessageID.IDS_FeatureAsyncMain, diagnostics); [](start = 24, length = 107)

It feels like we want to go through this code path even when the candidate is not async, but returns a task.